src/experimental/patterns/ec2-app.ts

```diff
@@ -0,0 +1,307 @@
+ *   The more an application is scaled, the longer it'll take to deploy.
+ * - It is not necessary to use the `autoscaling` Riff-Raff deployment type in conjunction with this.
+ *   Doing so won't break anything, however it'll lengthen the deployment time as instances will be rotated twice.
+ *   Please remove it from any manually created `riff-raff.yaml` file.
+ * - Your application should include a build number in its filename.
+ *   This value will change across builds, and therefore create a CloudFormation template difference to be deployed.
+ * - Requires the AWS CLI and `cfn-signal` binaries to be available on the instance, and on the `PATH`.
+ *   AMIgo adds these via the `aws-tools` role.
+ *
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html#cfn-attributes-updatepolicy-rollingupdate
+ * @see https://github.com/guardian/testing-asg-rolling-update
+ *
+ * @experimental
+ */
+export class GuEc2AppExperimental extends GuEc2App {
+  constructor(scope: GuStack, props: GuEc2AppExperimentalProps) {
+    const { minimumInstances, maximumInstances = minimumInstances * 2 } = props.scaling;
+    const { applicationPort } = props;
+    const { region, stackId } = scope;
+
+    super(scope, {
+      ...props,
+      updatePolicy: UpdatePolicy.rollingUpdate({
+        maxBatchSize: maximumInstances,
+        minInstancesInService: minimumInstances,
+        minSuccessPercentage: 100,
+        waitOnResourceSignals: true,
+
+        /*
+        If a scale-in event fires during an `AutoScalingRollingUpdate` operation, the update could fail and rollback.
+        For this reason, we suspend the `AlarmNotification` process, else availability of a service cannot be guaranteed.
+        Consequently, services cannot scale-out during deployments.
+        If AWS ever supports suspending scale-out and scale-in independently, we should allow scale-out.
+         */
+        suspendProcesses: [ScalingProcess.ALARM_NOTIFICATION],
+      }),
+    });
+
+    const { autoScalingGroup, targetGroup } = this;
+    const { userData, role } = autoScalingGroup;
+    const cfnAutoScalingGroup = autoScalingGroup.node.defaultChild as CfnAutoScalingGroup;
+
+    cfnAutoScalingGroup.desiredCapacity = minimumInstances.toString();
+
+    /*
+    Enable all ASG metrics for improved observability.
+    Specifying `granularity` without specifying `metrics` results in all metrics being enabled.
+    See https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-metricscollection.html
+     */
+    cfnAutoScalingGroup.metricsCollection = [
+      {
+        granularity: "1Minute",
+      },
+    ];
+
+    // TODO are these sensible values?
+    const signalTimeoutSeconds = Math.max(
+      targetGroup.healthCheck.timeout?.toSeconds() ?? 0,
+      cfnAutoScalingGroup.healthCheckGracePeriod ?? 0,
+      Duration.minutes(5).toSeconds(),
```

Comment:
I think it makes sense to explicitly couple this timeout to the healthcheck grace period (and I think this will be [2-minutes for most services](https://github.com/guardian/cdk/blob/12817108d384a561c85f7713a3e943c26fa3a6f3/src/patterns/ec2-app/base.ts#L428)).

Users can already tune this property based on how long their instances take to start passing healthchecks and adding this (potentially) longer and non-configurable default will cause the deployment of 'bad' builds to fail more slowly than necessary in many cases.

Reply:
If the above sounds sensible I think we could also drop a bit of logic and the `should create an ASG with the maximum resource signal timeout` unit test?

Reply:
This variable is used for the [`PauseTime` in the update policy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-updatepolicy.html#cfn-attributes-updatepolicy-rollingupdate-pausetime), which is described as:

> It determines the maximum time that CloudFormation waits to receive the required number of valid signals from the instances being replaced during a rolling update and from new instances being added by increasing the [DesiredCapacity](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-autoscaling-autoscalinggroup.html#cfn-autoscaling-autoscalinggroup-desiredcapacity) property of the AWS::AutoScaling::AutoScalingGroup resource.

My reading of this is the timeout starts when `DesiredCapacity` is changed, and thus includes the time spent [pending](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-lifecycle.html). Do we know if the healthcheck grace period covers pending too? I thought it doesn't, but can't find any docs to suggest either way.

Reply:
Done - https://github.com/guardian/cdk/pull/2417/commits/7a12f608297ced73a44c50b73c7312e22ffeafef.

This was a bit more involved than anticipated as we don't currently expose the ASG healthcheck grace period as a configurable property, it is hard-coded. Any [customisation is done via escape hatches](https://github.com/search?q=org%3Aguardian+.healthCheckGracePeriod+language%3ATypeScript&type=code&l=TypeScript). This means we need to use an `Aspect`.